rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp

```diff
@@ -184,22 +199,31 @@ class RMW_Connext_GuardCondition : public RMW_Connext_Condition
         DDS_GuardCondition_set_trigger_value(
           this->gcond, DDS_BOOLEAN_TRUE))
       {
-        RMW_CONNEXT_LOG_ERROR_SET("failed to trigger guard condition")
+        RMW_CONNEXT_LOG_ERROR_SET("failed to trigger internal guard condition")
         return RMW_RET_ERROR;
       }
 
       return RMW_RET_OK;
     }
 
-    std::lock_guard<std::mutex> lock(this->mutex_internal);
+    // If condition is attached to a waitset, lock its mutex to prevent
+    // modification of the `triggered_*` flags concurrently with WaitSet::wait().
+    std::mutex * waitset_mutex = this->waitset_mutex;
+    auto scope_exit = rcpputils::make_scope_exit(
+      [waitset_mutex]()
+      {
+        if (nullptr != waitset_mutex) {
+          waitset_mutex->unlock();
+        }
+      });
+    if (nullptr != waitset_mutex) {
+      waitset_mutex->lock();
+    }
```

Comment:
I think we must establish an order in which the WaitSet's mutex and a Condition's mutex are acquired, otherwise we risk ending up with a deadlock. So far, I was trying to enforce a locking scheme where the WaitSet's mutex is always acquired first, and then a condition's, because this simplifies the implementation of `WaitSet::wait()` and other private functions. It would be possible to invert the order, but then `WaitSet::on_condition_active()` would need to: unlock the WaitSet, lock a Condition, lock the WaitSet, then manipulate the condition, which I thought would be problematic for the whole "lost condition notifications" problem.

Maybe I'm overcomplicating this in my head, I'll review the code again and think more about it. Clearly the current solution is not correct, since there were several test failures in the latest CI.

Reply:
> I think we must establish an order in which the WaitSet's mutex and a Condition's mutex are acquired, otherwise we risk ending up with a deadlock

Yes, I imagined the order was inverted because of a deadlock, but I couldn't see what was it.

>  So far, I was trying to enforce a locking scheme where the WaitSet's mutex is always acquired first, and then a condition's

The problem with that, is that you need to synchronize with `attach()` and `detach()`, so you actually need to do it in the opposite order (if not, `waitset_mutex` and `condition_mutex` might be invalid).

> t would be possible to invert the order, but then WaitSet::on_condition_active() would need to: unlock the WaitSet, lock a Condition, lock the WaitSet, then manipulate the condition, which I thought would be problematic for the whole "lost condition notifications" problem.

Thanks for explaining.
I'm trying to see what is taking the condition `mutex_internal` in `WaitSet::on_condition_active()`, are those `has_data()`, `has_status()`, `has_triggered()`?
If those are the methods taking `mutex_internal`, they shouldn't.
See [this other comment](https://github.com/ros2/rmw_connextdds/pull/22#discussion_r607861543).

`WaitSet::on_condition_active()` is only called from `cv.wait(mutex, on_condition_active)`, i.e. it's called with the `waitset_mutex` hold.
No need to also take the `mutex_internal`, `waitset_mutex` is used to synchronize access to the conditions flags (`triggered_data`, `trigger_value`, etc) and `mutex_internal` is *only* used to synchronize with attach(), detach() (i.e. it's only prottecting the `waitset_mutex` and `condition_mutex` pointers, not the condition flags).

I hope that's clear, sorry if my comments are a bit confusing.

Reply:
Updated in [b35d211](https://github.com/ros2/rmw_connextdds/pull/22/commits/b35d2114ef3b072f7ba386d1406eff89756d2288)